### PR TITLE
Use Sync.Map for concurrency handling in Volume Cloning

### PR DESCRIPTION
**Pull request by madhuthorat**
Signed-off-by: Madhu Thorat <madhu.punjabi@in.ibm.com>

### CODE REVIEW

File: driver/csiplugin/controllerserver.go [hunk 1]
```diff
@@ -575,23 +575,30 @@ func (cs *ScaleControllerServer) CreateVolume(ctx context.Context, req *csi.Crea
...
+					Volume: &csi.Volume{
+						VolumeId:      volID,
+						CapacityBytes: int64(scaleVol.VolSize),
+						VolumeContext: req.GetParameters(),
+						ContentSource: volSrc,
+					},
+				}, nil
+			}
+		} else {
+			glog.Infof("volume: [%v] not found in volcopyjobstatusmap", scaleVol.VolName)
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:601]: Add the the verbose level to logs - glog.Infof --> glog.V(5).Infof

File: driver/csiplugin/controllerserver.go [hunk 2]
```diff
@@ -575,23 +575,30 @@ func (cs *ScaleControllerServer) CreateVolume(ctx context.Context, req *csi.Crea
...
+			volID := jobDetails.(VolCopyJobDetails).volID
+			glog.Infof("volume: [%v] found in volcopyjobstatusmap with volID: [%v], jobStatus: [%v]", scaleVol.VolName, volID, jobStatus)
+			if jobStatus == VOLCOPY_JOB_RUNNING {
+				glog.Errorf("volume:[%v] -  volume cloning request in progress.", scaleVol.VolName)
+				return nil, status.Error(codes.Aborted, fmt.Sprintf("volume cloning request in progress for volume: %s", scaleVol.VolName))
+			} else if jobStatus == VOLCOPY_JOB_FAILED {
+				glog.Errorf("volume:[%v] -  volume cloning job had failed", scaleVol.VolName)
+				return nil, status.Error(codes.Internal, fmt.Sprintf("volume cloning job had failed for volume:[%v]", scaleVol.VolName))
+                	} else if jobStatus == VOLCOPY_JOB_COMPLETED {
+				glog.Infof("volume:[%v] -  volume cloning request has already completed successfully.", scaleVol.VolName)
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:590]: Add the the verbose level to logs - glog.Infof --> glog.V(5).Infof

File: driver/csiplugin/controllerserver.go [hunk 3]
```diff
@@ -575,23 +575,30 @@ func (cs *ScaleControllerServer) CreateVolume(ctx context.Context, req *csi.Crea
...
-					CapacityBytes: int64(scaleVol.VolSize),
-					VolumeContext: req.GetParameters(),
-					ContentSource: volSrc,
-				},
-			}, nil
+		jobDetails, found := cs.Driver.volcopyjobstatusmap.Load(scaleVol.VolName)
+		if found {
+			jobStatus := jobDetails.(VolCopyJobDetails).jobStatus
+			volID := jobDetails.(VolCopyJobDetails).volID
+			glog.Infof("volume: [%v] found in volcopyjobstatusmap with volID: [%v], jobStatus: [%v]", scaleVol.VolName, volID, jobStatus)
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:582]: Add the the verbose level to logs - glog.Infof --> glog.V(5).Infof

File: driver/csiplugin/controllerserver.go (hunk 1 above)
Comment (madhuthorat) [reply driver/csiplugin/controllerserver.go]: Done

File: driver/csiplugin/controllerserver.go (hunk 2 above)
Comment (madhuthorat) [reply driver/csiplugin/controllerserver.go]: Done

File: driver/csiplugin/controllerserver.go (hunk 3 above)
Comment (madhuthorat) [reply driver/csiplugin/controllerserver.go]: Done

**Review by deeghuge — APPROVED** [review]

**Pull request merged**